src/instructlab/model/phased_training.py

```diff
@@ -0,0 +1,140 @@
+
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: list[pydantic.DirectoryPath]
+    finished_checkpoints: list[pydantic.DirectoryPath] = []
+    results: list[EvalResult] = []
+    best_checkpoint: EvalResult | None = None
+
+    @pydantic.field_serializer(
+        "checkpoints",
+    )
+    def pathlibPath_list_to_str(self, paths: list[pathlib.Path]) -> list[str]:
+        return [str(path) for path in paths]
+
+
+class TrainPhaseModel(pydantic.BaseModel):
+    """Stores info about training phase. Doesn't store training args because user can get those elsewhere, e.g. config."""
+
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: pydantic.DirectoryPath
+    # TODO: might want training args here, but can't currently do it
+    # without side-effects because model instance is changed inside of _training_phase
+
+
+class JournalModel(pydantic.BaseModel):
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    current_phase: TrainingPhases = TrainingPhases.TRAIN1
+    train_1: TrainPhaseModel | None = None
+    eval_1: EvalPhaseModel | None = None
+    train_2: TrainPhaseModel | None = None
+    eval_2: EvalPhaseModel | None = None
+    final_output: EvalResult | None = None
+
+
+class TrainingJournal:
+    def __init__(self, journalfile: pathlib.Path):
+        self.journalfile = journalfile
+        if journalfile.is_file():
+            if journalfile.stat().st_size == 0:
+                logger.debug("Journalfile is empty- creating a new journalfile.")
+                # if journal is empty, use path but initialize model.
+                self.journal = JournalModel()
+                self.commit()
+            else:
+                logger.debug("Journalfile exists- trying to load.")
+                with open(journalfile, "r") as f:
+                    model_dict = yaml.safe_load(stream=f)
+                    try:
+                        self.journal = JournalModel.parse_obj(model_dict)
+                        logger.debug("Model loaded from Journalfile")
+                    except pydantic.ValidationError:
+                        print(
+                            "Tried to load Journal from Journalfile but encountered an error. Journalfiles must be valid or empty."
+                        )
+                        raise
+        elif journalfile.is_dir():
+            # add 'journalfile.yaml' to the directory by convention
+            logger.debug(
+                "'journalfile' path points to a directory. Creating a new 'journalfile.yaml'"
+            )
+            self.journalfile = journalfile / "journalfile.yaml"
+            self.journal = JournalModel()
+            self.commit()
+        else:
+            raise FileExistsError(f"'journalfile' doesn't exist: {str(journalfile)}")
```

Comment:
I prefer to fail early whenever possible, and this seems like a good opportunity for that.​

Reply:
I fixed this up, it has (imo) better behavior now.

If file doesn't exist, it just errors out. If it does exist, but is empty, it initializes it fresh. If it does exist, it tries to parse/load. If parsing/loading fails, it breaks.